nocturne/envs/vec_env_ma.py

```diff
@@ -0,0 +1,242 @@
+        }
+
+        # Take a step to obtain dicts
+        next_obses_dict, rew_dict, done_dict, info_dict = self.env.step(agent_actions)
+
+        # Update dead agents based on most recent done_dict
+        for agent_id, is_done in done_dict.items():
+            if is_done and agent_id not in self.dead_agent_ids:
+                self.dead_agent_ids.append(agent_id)
+                # Store agents' last info dict
+                self.last_info_dicts[agent_id] = info_dict[agent_id].copy()
+
+        # Convert dicts to arrays
+        obs_all = np.full(
+            fill_value=np.nan,
+            shape=(self.num_envs, self.env.observation_space.shape[0]),
+        )
+        rew_all = np.full(fill_value=np.nan, shape=(self.num_envs))
+        done_all = np.full(fill_value=np.nan, shape=(self.num_envs))
+        info_all = []
+
+        for idx, key in enumerate(self.agent_ids):
+            # Store data if available; otherwise leave as NaN
+            if key in next_obses_dict:
+                obs_all[idx, :] = next_obses_dict[key]
+                rew_all[idx] = rew_dict[key]
+                done_all[idx] = done_dict[key] * 1  # Will be 0 or 1 if valid, NaN otherwise
+
+        # OVERRIDE old buffer vals with with new ones (for all envs)
+        for env_idx in range(self.num_envs):
+            info_all.append(info_dict[key])
+            self.buf_rews[env_idx] = rew_all[env_idx]
+            self.buf_dones[env_idx] = done_all[env_idx]
+            self.buf_infos[env_idx] = info_all[env_idx]
+
+        # Save step reward obtained across all agents
+        self.rewards.append(sum(rew_dict.values()))
+        self.agents_in_scene.append(len(self.agent_ids))
+
+        # O(t) = O(t+1)
+        self._save_obs(obs_all)
+
+        # Reset episode if ALL agents are done
+        if done_dict["__all__"]:
+            for agent_id in self.agent_ids:
+                # Store total number of collisions and goal achievements across rollout
+                self.ep_collisions += self.last_info_dicts[agent_id]["collided"] * 1
+                self.ep_goal_achived += self.last_info_dicts[agent_id]["goal_achieved"] * 1
+
+            # Store the fraction of agents that collided in episode
+            self.frac_collided.append(self.ep_collisions / len(self.agent_ids))
+            self.frac_goal_achieved.append(self.ep_goal_achived / len(self.agent_ids))
+
+            # Save final observation where user can get it, then reset
+            for env_idx in range(len(self.agent_ids)):
+                self.buf_infos[env_idx]["terminal_observation"] = obs_all[env_idx]
+
+            # Log episode stats
+            ep_len = self.step_num
+            self.n_episodes += 1
+            self.episode_lengths.append(ep_len)
+
+            # Reset
+            obs_all = self.reset()
+
+        return (
+            self._obs_from_buf(),
+            np.copy(self.buf_rews),
+            np.copy(self.buf_dones),
+            deepcopy(self.buf_infos),
```

Comment:
I would check if this copy is necessary, I suspect on the sb3 end there's another copy that's happening